v2/brokers.go

```diff
@@ -0,0 +1,232 @@
+package liftbridge
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"sync"
+
+	proto "github.com/liftbridge-io/liftbridge-api/go"
+	"github.com/serialx/hashring"
+	"google.golang.org/grpc"
+)
+
+type ackReceivedFunc func(*proto.PublishResponse)
+
+// brokers represent a collection of connections to brokers.
+type brokers struct {
+	mu          sync.RWMutex
+	brokers     []*broker
+	brokersMap  map[string]*broker
+	ring        *hashring.HashRing
+	opts        []grpc.DialOption
+	ackReceived ackReceivedFunc
+}
+
+func newBrokers(ctx context.Context, addrs []string, opts []grpc.DialOption, ackReceived ackReceivedFunc) (*brokers, error) {
+	brokersSlice := make([]*broker, 0, len(addrs))
+	brokersMap := make(map[string]*broker, len(addrs))
+
+	// Ensure all connections are closed if an error occurs.
+	connectedBrokers := make([]*broker, 0, len(addrs))
```

Comment:
yeah interesting question on what we should do on a partial success. I don't think we'll want to prevent publishing if a broker is not contactable. It might end up being better to succeed if a connection to any broker succeeds

Reply:
Also long run the publication streams should probably be created on demand but I don't think you need to do that now

Reply:
> yeah interesting question on what we should do on a partial success. I don't think we'll want to prevent publishing if a broker is not contactable. It might end up being better to succeed if a connection to any broker succeeds

The current code initiates two connections and closes both of them in case one of them fails, but yeah let's see what Tyler says.

> Also in the long run the publication streams should probably be created on demand but I don't think you need to do that now

That could be an option maybe? Connecting might be slow, so establishing and maintaining an initial connection could make sense depending on the use case.